### PR TITLE
fix(viewer): apply geoid correction by default for orthometric IFC heights (#1355)

### LINKED ISSUE #1355: Model on map not showing properly
The model has been used in production and it is placed correctly. Do you know why it looks like this? Model attached in the other issue

Image

### PR DESCRIPTION
**Pull request by louistrue**
## Refreshed against main (2026-07-10)

Merged `origin/main` into the branch (branch was 162 commits behind). The georef/Cesium code on main had moved, so:

- Resolved one conflict in `apps/viewer/src/lib/geo/reproject.test.ts` (purely additive: main added `reprojectPointToLatLon` / `reprojectionInputKey` suites, this branch added the `computeCesiumModelOrigin geoid correction default (#1355)` suite - both kept).
- Reverted an unrelated stray `AGENTS.md` edit that was sitting uncommitted in the worktree.
- No changes to the fix's logic were required by the merge - the geoid path in `cesium-bridge.ts` / `cesium-placement.ts` still applies the EGM96 correction by default, gated solely on `heightsAreEllipsoidal`.

### Verification
- Viewer `tsc --noEmit`: clean (after rebuilding the stale `@ifc-lite/diff`, `@ifc-lite/clash`, `@ifc-lite/bcf`, `@ifc-lite/data` dists in the worktree).
- Geo test suites green: `apps/viewer` `tsx --test` over `src/lib/geo/**` - **151/151 pass**, including the new geoid-default coverage and the merged reproject suites.

### On the EPSG:5709 / vertical-datum discussion (#1355)
MennoMekes' workaround was to set `EPSG:5709` (NAP) as the vertical datum so the model wasn't buried. Under the old code the geoid correction was gated on a declared `VerticalDatum`, so setting it is what triggered the +N lift. This PR makes the correction the default for every georeferenced model, so that manual workaround is no longer needed.

There is **no double-correction risk**: model placement adds `egm96Undulation(lat, lon)` exactly once, gated only by `heightsAreEllipsoidal`. `verticalDatum` never enters the altitude math (it only influences which terrain DEM is preferred for the camera/clip hint via `shouldPreferOrthometricTerrain`). So a file that still carries `EPSG:5709` is lifted by N once, not twice.

---

## Summary

Georeferenced models sat ~40-50 m below the Cesium world terrain. `IfcMapConversion.OrthogonalHeight` is orthometric (height above the vertical datum / mean sea level), while Cesium positions geometry by ellipsoidal height (above the WGS84 ellipsoid). The two differ by the EGM96 geoid undulation N (~+43..49 m in central Europe), so without the correction the model sinks ~N below terrain.

The EGM96 correction already existed (merged #1376) but was gated on a declared `IfcProjectedCRS.VerticalDatum`, which is optional in IFC4/IFC4X3 and routinely omitted by exporters (very common for Dutch RD-New / NAP files). When absent, the correction silently no-op'd and the model stayed buried.

## Fix
- Decouple the two concerns that were fused on the `verticalDatum` flag:
  - `shouldPreferOrthometricTerrain()` (terrain-sampling heuristic) stays gated on a declared vertical datum.
  - `shouldApplyGeoidUndulation()` is new and defaults ON for every georeferenced model, because the authored altitude is orthometric by spec regardless of whether `VerticalDatum` is declared.
- Add an opt-out for the rare file whose `OrthogonalHeight` is genuinely ellipsoidal: `cesiumHeightsAreEllipsoidal` (default false), surfaced as a toggle in the Georeferencing panel.
- Keep the below-terrain clip / terrain frame consistent (`cesium-bridge.ts`, `CesiumOverlay.tsx`).

## Verification
- `apps/viewer` `tsx --test` geo suite - 36/36 pass, including a new "geoid correction default (#1355)" suite (correction applied by default; skipped when `heightsAreEllipsoidal`) plus `egm96-undulation` and `reproject` coverage.
- Visual confirmation against live Cesium terrain is deferred to manual review (terrain rendering is not unit-testable here).

No changeset (apps/viewer is unpublished). Conventional commit, no co-author trailer.

## Residual risk
Broadening the default could over-correct a file whose heights are already ellipsoidal; mitigated by the `heightsAreEllipsoidal` opt-out and the spec basis (IFC altitude is orthometric by default). EGM96 on its grid is sub-metre to ~1-2 m accurate - right for gross placement, not survey grade.

Closes #1355

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a new option to treat Cesium heights as ellipsoidal, with updated snapping behavior to match the selected height frame.
  * Improved terrain sampling so saved heights can align more accurately with the displayed terrain surface.

* **Bug Fixes**
  * Fixed height conversion when setting orthogonal/map conversion heights from sampled terrain.
  * Reset height-related Cesium settings when loading a new file to prevent old values from carrying over.

* **Tests**
  * Expanded coverage for geoid correction, terrain snapping, and height conversion edge cases.

### CODE REVIEW

File: apps/viewer/src/lib/geo/cesium-bridge.ts [hunk 1]
```diff
@@ -119,10 +126,11 @@ export async function computeCesiumModelOrigin(
...
-    // unambiguous "these heights are orthometric" signal; models that don't
-    // declare one are left exactly as before. (#1355)
-    const geoidUndulation = shouldPreferOrthometricTerrain(projectedCRS)
+    // +49 m in Switzerland). This is the DEFAULT for every georeferenced model
+    // — the authored altitude is orthometric by spec, whether or not a
+    // VerticalDatum is declared (it is optional and routinely omitted). The
+    // only opt-out is a file whose heights are already ellipsoidal. (#1355)
+    const geoidUndulation = shouldApplyGeoidUndulation(heightsAreEllipsoidal)
       ? egm96Undulation(lat, lon)
       : 0;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/lib/geo/cesium-bridge.ts:154]: Confirmed, valid finding. The core fix (default geoid correction on auto-placement) is correct and unaffected; this is specifically the manual "Set OrthogonalHeight to sampled terrain height" button. Diagnosis: `cesiumTerrainHeight` stores the raw sample whose frame depends on source (Open-Meteo = orthometric, Cesium getHeight/sampleHeight = ellipsoidal), and with the read path now adding N, an ellipsoidal sample saved as OrthogonalHeight reads back N too high. The correct fix is to persist an orthometric target: `OrthogonalHeight_target = terrainHForFrame - modelOrigin.geoidUndulation` (verified consistent across orthometric/ellipsoidal sources and the heightsAreEllipsoidal opt-out). Since this is a georef-frame round-trip that warrants visual confirmation in live Cesium, I'm flagging it for the maintainer rather than shipping it blind in this PR; will follow up.

[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/lib/geo/cesium-bridge.ts:154]: Implemented in fd14f883. Added `orthometricTargetForTerrain(ellipsoidalTerrain, N) = ellipsoidalTerrain - N` plus a `cesiumTerrainSaveHeight` store field computed in CesiumOverlay (where the terrain frame and the applied undulation are known: terrainHForFrame minus the same N the bridge applied, 0 when heights are flagged ellipsoidal). Both snap buttons now persist that value, so saving a Cesium-sourced (ellipsoidal) sample no longer reads back +N too high; the displayed sampled value is unchanged. Verified across orthometric/ellipsoidal sources and the opt-out, with a round-trip unit test. Full visual confirmation in live Cesium still recommended before merge.

File: apps/viewer/src/components/viewer/properties/GeoreferencingPanel.tsx [hunk 2]
```diff
@@ -782,14 +787,33 @@ export function GeoreferencingPanel({ georef, modelId, enableEditing, schemaVers
           {cesiumTerrainHeight !== null && editable && modelId && (
             <div className="flex items-center gap-1 ml-5">
               <button
-                onClick={() => handleSave('mapConversion', 'orthogonalHeight', oHeightForBaseAltitude(cesiumTerrainHeight))}
+                onClick={() => handleSave('mapConversion', 'orthogonalHeight', oHeightForBaseAltitude(cesiumTerrainSaveHeight ?? cesiumTerrainHeight))}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/properties/GeoreferencingPanel.tsx]: Fixed in dbb79875. Both snap buttons now gate on the corrected `cesiumTerrainSaveHeight` and no longer fall back to the raw ellipsoidal sample, so an uncorrected value can never be persisted; the button simply hides until the corrected target is ready.

File: apps/viewer/src/store/slices/cesiumSlice.ts [hunk 3]
```diff
@@ -208,12 +231,15 @@ export const createCesiumSlice: StateCreator<CesiumSlice & CesiumCrossSliceState
...
       cesiumTerrainSource: null,
+      cesiumTerrainSaveHeight: null,
       cesiumTerrainClipY: null,
     });
   },
   setCesiumTerrainHeight: (height) => set({ cesiumTerrainHeight: height }),
   setCesiumTerrainSource: (source) => set({ cesiumTerrainSource: source }),
+  setCesiumTerrainSaveHeight: (height) => set({ cesiumTerrainSaveHeight: height }),
   setCesiumSourceModelId: (modelId) => set({ cesiumSourceModelId: modelId }),
+  setCesiumHeightsAreEllipsoidal: (ellipsoidal) => set({ cesiumHeightsAreEllipsoidal: ellipsoidal }),
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/cesiumSlice.ts]: Fixed in dbb79875. `setCesiumHeightsAreEllipsoidal` now clears `cesiumTerrainSaveHeight`, so toggling the geoid frame invalidates the stale target; the overlay effect (which depends on heightsAreEllipsoidal) recomputes it under the new frame before the next snap.

**Comment by louistrue** [comment]
Addressed the remaining CodeRabbit outside-diff finding in e279f4e3: resetViewerState now clears cesiumTerrainSaveHeight alongside cesiumTerrainHeight, so a new file cannot reuse the previous file's snap target. Also merged main (a51f3a1a, no conflicts; no file overlap with the branch changes). Viewer typecheck, the geo/cesium unit tests, and the full viewer suite (1254 pass / 0 fail) are green locally.

**Comment by louistrue** [comment]
Brought the branch up to date with current main (merge commit e4cd08c4). The only conflict was `apps/viewer/src/lib/geo/reproject.test.ts`, where both sides had added independent test suites; resolved as a purely additive union, keeping every test from both sides (main's reprojectPointToLatLon / reprojectionInputKey suites from #1679 plus the KMZ-related helper tests from #1430, and this branch's geoid-correction-default suite for #1355). All production files auto-merged.

Verification on the merged branch:
- viewer typecheck: green
- merged `reproject.test.ts`: 20/20 pass across 5 suites
- full viewer test suite: 1543 pass, 0 fail, 2 skipped

One verification step remains open before merge: the manual Cesium terrain visual check against the #1355 reporter models. A commenter noted that the height correction alone may not fully fix placement compared to geo.buildingsmart.nl, so it is worth eyeballing the terrain placement on those models.

**Comment by louistrue** [comment]
@coderabbitai review
